### PR TITLE
[dynamo] Fix and simplify hanlding of `Set.update` method

### LINKED ISSUE #141283: `torch.compile` fails with internal error on `Set.update` with iterator input
### 🐛 Describe the bug

This was exposed during user empathy day, see [notes](https://github.com/StrongerXi/bark/blob/torch-compile/notes.md#internaltorchdynamoerror-setvariable).

Repro:
```python
@torch.compile(backend="eager")
def run(x, int_set, ns):
    int_set.update(map(int, ns))
    return x + 1
int_set = set()
ns = [1, 2, 1]
run(torch.ones(1), int_set, ns)
```

### Error logs

```
Traceback (most recent call last):
  File "/Users/ryanguo99/Documents/work/scratch/test.py", line 30, in <module>
    test()
  File "/Users/ryanguo99/Documents/work/scratch/test.py", line 26, in test
    run(torch.ones(1), int_set, ns)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/eval_frame.py", line 556, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 1403, in __call__
    return self._torchdynamo_orig_callable(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 1187, in __call__
    result = self._inner_convert(
             ^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 549, in __call__
    return _compile(
           ^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 1034, in _compile
    raise InternalTorchDynamoError(
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 984, in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 712, in compile_inner
    return _compile_inner(code, one_graph, hooks, transform)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_utils_internal.py", line 95, in wrapper_function
    return function(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 747, in _compile_inner
    out_code = transform_code_object(code, transform)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/bytecode_transformation.py", line 1348, in transform_code_object
    transformations(instructions, code_options)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 233, in _fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/convert_frame.py", line 664, in transform
    tracer.run()
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 2841, in run
    super().run()
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 1032, in run
    while self.step():
          ^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 944, in step
    self.dispatch_table[inst.opcode](self, inst)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 641, in wrapper
    return inner_fn(self, inst)
           ^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 2314, in CALL
    self._call(inst)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 2308, in _call
    self.call_function(fn, args, kwargs)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/symbolic_convert.py", line 879, in call_function
    self.push(fn.call_function(self, args, kwargs))  # type: ignore[arg-type]
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/misc.py", line 1032, in call_function
    return self.obj.call_method(tx, self.name, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/dicts.py", line 570, in call_method
    return super().call_method(tx, name, args, kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/dicts.py", line 357, in call_method
    dict_vt = BuiltinVariable.call_custom_dict(tx, dict, args[0])
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/builtin.py", line 1388, in call_custom_dict
    items = dict(
            ^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/builtin.py", line 1389, in <genexpr>
    x.force_unpack_var_sequence(tx)
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/base.py", line 380, in force_unpack_var_sequence
    return self.unpack_var_sequence(tx)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryanguo99/Documents/work/pytorch/torch/_dynamo/variables/constant.py", line 112, in unpack_var_sequence
    raise NotImplementedError from e
torch._dynamo.exc.InternalTorchDynamoError: NotImplementedError:

from user code:
   File "/Users/ryanguo99/Documents/work/scratch/test.py", line 22, in run
    int_set.update(map(int, ns))
```

### Versions

7ced49d2ccf, Python 3.12

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @amjames

### PR DESCRIPTION
**Pull request by StrongerXi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141286

The old implementation of `SetVariable.call_method("update", ...)` was
incorrectly becacuse it wouldn't handle iterable inputs. This patches
removes the input type restriction altogether, and implements the method
as a polyfill (like how most of the other set methods are handled).

Fixes #141283.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: test/dynamo/test_misc.py [hunk 1]
```diff
@@ -4834,6 +4834,19 @@ def fn(y):
         self.assertEqual(opt_fn(x), x)
         self.assertEqual(cnts.op_count, 1)
 
+    def test_set_update(self):
+        @torch.compile(backend="eager")
```
Comment (anijain2305) [inline test/dynamo/test_misc.py:4838]: fullgraph=True

**Review by anijain2305 — APPROVED** [review]

File: test/dynamo/test_misc.py (hunk 1 above)
Comment (StrongerXi) [reply test/dynamo/test_misc.py:4838]: Good call, updated.

**Comment by StrongerXi** [comment]
Rebase.

**Pull request closed without merging**